Fix TabulatedFluidProperties for multiphase class PorousFlowWaterVapor

aikubo opened the pull request:
closes #27457

aikubo committed:
Add functions so TabulatedFluidProperties works with multiphase class PorousFlowWaterVapor

aikubo committed:
add tests in PorousFlow fluidstate

remove debug statements

style patch

remove white space #27457

Add FluidPropertiesForwardError, fix a few formatting issues, fix forwarding of vapor_temperature_ad

remove _ad versions

Update modules/fluid_properties/src/fluidproperties/TabulatedFluidProperties.C

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Update SinglePhaseFluidProperties.h

Apply suggestions from code review

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

style patch

remove _ad for vapor temperature and T_from_p_h

aikubo commented:
This is my first moose pull request so I expect there to be some formatting or other errors. 
thanks for your patience

Also, I found that using the VScode git interface to try and push and pull causes errors. So I recommend other newbies do it all on terminal.

GiudGiud commented:
so there's these two errors which should be fairly easy to fix

```
##########################################################################
ERROR: Your patch does not contain a valid ticket reference! (i.e. #1234)
Merge branch 'tfp_water_vapor' of https://github.com/aikubo/moose into test
style patch
remove debug statements
add tests in PorousFlow fluidstate
Add functions so TabulatedFluidProperties works with multiphase class PorousFlowWaterVapor

ERROR: The following files contain trailing whitespace after applying your patch:
	modules/porous_flow/test/tests/fluidstate/water_vapor_tab.i

Run the "delete_trailing_whitespace.sh" script in your $MOOSE_DIR/scripts directory.
##########################################################################
```

GiudGiud commented on modules/fluid_properties/include/fluidproperties/TabulatedFluidProperties.h:
```diff
@@ -110,6 +112,23 @@ class TabulatedFluidProperties : public SinglePhaseFluidProperties
 
   virtual void vaporPressure(Real temperature, Real & psat, Real & dpsat_dT) const override;
 
+  template <typename T>
+  void vaporPressureTemplate(const T & temperature, T & psat, T & dpsat_dT) const;
```
is this defined?

aikubo replied:
Again, I was trying to keep it consistent with Water97 but I did forget to define it so I can just take it out.

GiudGiud replied:
Sounds good. Let s remove that one

aikubo replied:
removed.

GiudGiud commented on modules/fluid_properties/src/fluidproperties/TabulatedFluidProperties.C:
```diff
@@ -76,6 +76,10 @@ TabulatedFluidProperties::validParams()
+  param.addParam<bool>("p_h_variables",
+                       false,
+                       "If using pressure enthalpy variables, interpolates from PT to h"
+                       "may be wrong especially in multiphase.");
```
say what it does?

aikubo replied:
addressed in [a6a0e70](https://github.com/idaholab/moose/pull/27506/commits/a6a0e70fd320f10a2453e722ddc0c69452dc09be)

GiudGiud commented on modules/fluid_properties/src/fluidproperties/TabulatedFluidProperties.C:
```diff
@@ -609,6 +623,15 @@ TabulatedFluidProperties::h_from_p_T(Real pressure, Real temperature) const
+  if (_fp) // Assuming _fp can handle ADReal types
+    return _fp->h_from_p_T(pressure, temperature);
+  else
+    paramError("fp", "Fluid properties object does not support ADReal computations.");
```
cant handle this specific one

maybe create a routine with a comprehensive error message, that says what is missing
TFP::fluidPropertyForwardError(std::string desired_routine, bool is_ad)

aikubo replied:
added in [a6a0e70](https://github.com/idaholab/moose/pull/27506/commits/a6a0e70fd320f10a2453e722ddc0c69452dc09be)

GiudGiud commented on modules/fluid_properties/src/fluidproperties/TabulatedFluidProperties.C:
```diff
@@ -895,6 +918,100 @@ TabulatedFluidProperties::vaporPressure(Real temperature, Real & psat, Real & dp
+
+  const FPDualReal p = pressure;
+
+  return vaporTemperature_ad(p).value();
```
use the same forward as the others instead?

aikubo replied:
addressed in [a6a0e70](https://github.com/idaholab/moose/pull/27506/commits/a6a0e70fd320f10a2453e722ddc0c69452dc09be)

GiudGiud commented on modules/fluid_properties/src/fluidproperties/TabulatedFluidProperties.C:
```diff
@@ -895,6 +918,100 @@ TabulatedFluidProperties::vaporPressure(Real temperature, Real & psat, Real & dp
+  if (_fp)
+    _fp->vaporTemperature(pressure, Tsat, dTsat_dp);
+  else
+    mooseError("vaporTemperature not specified.");
```
It's not really that it s not specified. It s more:

"Tabulation does not contain the vapor temperature. To use the TabulatedFluidProperties and compute the vaporTemperature, please provide another fluid property using the 'fp' parameter, which will be used to compute the vapor temperature"

and same, make it a standard error message in a TFP routine, which you can re-use for each of these routines

GiudGiud commented on modules/fluid_properties/src/fluidproperties/TabulatedFluidProperties.C:
```diff
@@ -1272,20 +1402,19 @@ TabulatedFluidProperties::checkInputVariables(T & pressure, T & temperature) con
-      throw MooseException("Pressure " + Moose::stringify(pressure) +
+      throw MooseException("Pressure " + 
                            " is outside the range of tabulated pressure (" +
-                           Moose::stringify(_pressure_min) + ", " +
-                           Moose::stringify(_pressure_max) + ").");
+                           
+                           
```
revert

GiudGiud commented on modules/fluid_properties/src/fluidproperties/TabulatedFluidProperties.C:
```diff
@@ -1272,20 +1402,19 @@ TabulatedFluidProperties::checkInputVariables(T & pressure, T & temperature) con
-                 " is outside the range of tabulated temperature (" +
-                 Moose::stringify(_temperature_min) + ", " + Moose::stringify(_temperature_max) +
+      mooseError("Temperature " + " is outside the range of tabulated temperature (" +
+
```
same

GiudGiud commented on modules/fluid_properties/src/fluidproperties/TabulatedFluidProperties.C:
```diff
@@ -1299,17 +1428,17 @@ TabulatedFluidProperties::checkInitialGuess() const
                    " is outside the range of tabulated "
                    "temperature (" +
-                   Moose::stringify(_temperature_min) + ", " + Moose::stringify(_temperature_max) +
+                   
```
same

GiudGiud commented on modules/porous_flow/test/tests/fluidstate/tests:
```diff
@@ -181,6 +181,15 @@
+    input = 'water_vapor_tab.i'
+    csvdiff = 'water_vapor_twophase.csv'
+    threading = '!pthreads'
+    requirement = 'The system shall be able to use tabulate fluid properties.'
```
```suggestion
    requirement = 'The system shall be able to use tabulated fluid properties in conjunction with another fluid properties object to provide non-tabulated properties.'
```

GiudGiud replied:
let s have the requirement be a little more specific

GiudGiud reviewed (changes requested):
if it becomes too complicated we will want to use a tabulation for all these things instead

aikubo commented on modules/fluid_properties/include/fluidproperties/SinglePhaseFluidProperties.h:
```diff
@@ -311,8 +311,20 @@ class SinglePhaseFluidProperties : public FluidProperties
   virtual void vaporTemperature(Real p, Real & Tsat, Real & dTsat_dp) const;
-  DualReal vaporTemperature(const DualReal & p) const;
+  virtual DualReal vaporTemperature(const DualReal & p) const;
+  virtual FPDualReal vaporTemperature_ad(const FPDualReal & p) const;
```
`virtual FPDualReal vaporTemperature_ad(const FPDualReal & p) const;`
is there to match the declaration in Water97FluidProperties

https://github.com/idaholab/moose/blob/1145d861a4cf31c9a159563694bd7d5619353a09/modules/fluid_properties/include/fluidproperties/Water97FluidProperties.h#L813

Maybe it's not necessary but I was trying to keep it consistent. There are some inconsistencies between AD versions and non-AD versions throughout Water97 and other fluid properties, some use template functions which I think is the prefered implementation and others use the `_ad` implementation.

aikubo replied:
In Water97:

The general format is that there is a Real version, an ADReal version, and then a void version which takes the derivatives and then another void version which is AD. All of these reference two template functions one with the derivatives and one without.

https://github.com/idaholab/moose/blob/1145d861a4cf31c9a159563694bd7d5619353a09/modules/fluid_properties/src/fluidproperties/Water97FluidProperties.C#L80C1-L107C2

However, for some of the functions (vaporTemperature and T_from_p_h) there's no template function just a `_ad` version which is used for each of the functions. 
https://github.com/idaholab/moose/blob/1145d861a4cf31c9a159563694bd7d5619353a09/modules/fluid_properties/src/fluidproperties/Water97FluidProperties.C#L511C1-L549C2

I think both implementations have the same behavior so I'm not totally sure what the difference is. Also, I'm still not clear on the void versions with the derivatives but they are necessary for the functioning of things like PorousFlowWaterVapor. When I looked into what PFWaterVapor calls it calls `void Water97FluidProperties::vaporTemperature` then  `Water97FluidProperties::vaporTemperature_ad` in 

https://github.com/idaholab/moose/blob/087d66eb2aace3e6b0fe1db7b316749254c301d2/modules/porous_flow/src/userobjects/PorousFlowWaterVapor.C#L105C1-L109C49

cpgr replied:
From memory (it was a long time ago!) those `foo_ad` methods are really just used internally to avoid hand-coding the derivatives. I think the `FPDualReal` ones use a smaller derivative container for speed. None of these are exposed in the public `SinglePhaseFluidProperties` API, so you probably don't want them in TabulatedFluidProperties.

I did have an idea which I'll add in a comment below

aikubo replied:
Removed in d2f638c

GiudGiud commented:
let's remove the merge commit. We prefer rebase. You can remove it by:
```
git reset --hard a6a0e70
git fetch upstream
git rebase upstream/devel
git push origin HEAD:tfp_water_vapor -f
```

you have more formatting issues
```
##########################################################################
ERROR: The following files contain trailing whitespace after applying your patch:
	modules/porous_flow/test/tests/fluidstate/water_vapor_tab.i

Run the "delete_trailing_whitespace.sh" script in your $MOOSE_DIR/scripts directory.
##########################################################################
```

if you are using VSCode you need to use our user settings, and it wont be a problem ever again
https://mooseframework.inl.gov/help/development/VSCode.html

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2207484/) on a6a0e70 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27506/site)

This comment will be updated on new commits.

GiudGiud commented on modules/fluid_properties/src/fluidproperties/TabulatedFluidProperties.C:
```diff
@@ -892,7 +917,88 @@ TabulatedFluidProperties::vaporPressure(Real temperature, Real & psat, Real & dp
+{
+
+  if (_fp)
+  {
+    return _fp->T_from_p_h(pressure, enthalpy);
+  }
```
```suggestion
    return _fp->T_from_p_h(pressure, enthalpy);
```

GiudGiud reviewed (changes requested):
minor things

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2219057/) on 843d574 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27506/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27506/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b3efaa2fb20af48d2c07a376c6c0f653ae8a66db`

GiudGiud commented:
cleaned up the commit history a little. let's see if tests pass

aikubo commented:
@cpgr @GiudGiud 

I went back to my original input file from #27159. And I get a different error using tabulated fluid properties than from the test input file I made `water_vapor_tab.i`

```
*** ERROR ***
/home/akh/myprojects/moose_projects/dikes/sims/multi/MVE_s_from_h_p_error.i:186.5:
The following error occurred in the FluidProperties 'water97' of type Water97FluidProperties.

The fluid properties class 'Water97FluidProperties' has not implemented the method below, which computes derivatives of fluid properties with regards to the flow variables. If your application requires this method, you must either implement it or use a different fluid properties  class.

virtual void SinglePhaseFluidProperties::s_from_h_p(libMesh::Real, libMesh::Real, libMesh::Real&, libMesh::Real&, libMesh::Real&) const

You can avoid this error by neglecting the unimplemented derivatives of fluid properties by setting the 'allow_imperfect_jacobians' parameter

Stack frames: 25
0: libMesh::print_trace(std::ostream&)
1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
2: MooseBase::callMooseError(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool) const
3: /home/akh/myprojects/moose_projects/moose/modules/fluid_properties/lib/libfluid_properties-opt.so.0(+0xbbd84) [0x7f0b716bbd84]
4: /home/akh/myprojects/moose_projects/moose/modules/fluid_properties/lib/libfluid_properties-opt.so.0(+0x13fce4) [0x7f0b7173fce4]
5: /home/akh/myprojects/moose_projects/moose/modules/fluid_properties/lib/libfluid_properties-opt.so.0(+0x143574) [0x7f0b71743574]
6: SinglePhaseFluidProperties::T_from_p_h(double, double, double&, double&, double&) const
7: /home/akh/myprojects/moose_projects/moose/modules/fluid_properties/lib/libfluid_properties-opt.so.0(+0x126f35) [0x7f0b71726f35]
8: PorousFlowWaterVapor::thermophysicalProperties(MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<64ul> >, true> const&, MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<64ul> >, true> const&, unsigned int, FluidStatePhaseEnum&, std::vector<FluidStateProperties, std::allocator<FluidStateProperties> >&) const
9: PorousFlowWaterVapor::thermophysicalProperties(double, double, unsigned int, std::vector<FluidStateProperties, std::allocator<FluidStateProperties> >&) const
10: PorousFlowFluidStateBaseMaterialTempl<false>::initQpStatefulProperties()
11: PorousFlowFluidStateSingleComponentTempl<false>::initQpStatefulProperties()
12: MaterialBase::initStatefulProperties(unsigned int)
13: MaterialPropertyStorage::initStatefulProps(unsigned int, std::vector<std::shared_ptr<MaterialBase>, std::allocator<std::shared_ptr<MaterialBase> > > const&, unsigned int, libMesh::Elem const&, unsigned int)
14: ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator()(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool)
15: /home/akh/myprojects/moose_projects/moose/framework/libmoose-opt.so.0(+0x20e16d0) [0x7f0b74ee16d0]
16: FEProblemBase::initElementStatefulProps(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool)
17: FEProblemBase::initialSetup()
18: Transient::init()
19: MooseApp::executeExecutioner()
20: MooseApp::run()
21: main
22: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f0b68c29d90]
23: __libc_start_main
24: ../../dikes-opt(+0x33ef) [0x557abb8ce3ef]
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor

```

This is an issue with the initial condition for enthalpy. Basically if you're not in the two phase zone, it gives this error. But I don't know why because I don't think any of these objects ever call `s_from_h_p` and it ONLY happens when use Tabulated Fluid Properties. 

Here's a MVE:

```
[Mesh]
    [gen]
      type = GeneratedMeshGenerator
      dim = 2

    []
  []

  [GlobalParams]
    PorousFlowDictator = dictator
  []

  [AuxVariables]
    [pressure_gas]
      order = CONSTANT
      family = MONOMIAL
    []
    [pressure_water]
      order = CONSTANT
      family = MONOMIAL
    []
    [enthalpy_gas]
      order = CONSTANT
      family = MONOMIAL
    []
    [enthalpy_water]
      order = CONSTANT
      family = MONOMIAL
    []
    [saturation_gas]
      order = CONSTANT
      family = MONOMIAL
    []
    [saturation_water]
      order = CONSTANT
      family = MONOMIAL
    []
    [density_water]
      order = CONSTANT
      family = MONOMIAL
    []
    [density_gas]
      order = CONSTANT
      family = MONOMIAL
    []
    [viscosity_water]
      order = CONSTANT
      family = MONOMIAL
    []
    [viscosity_gas]
      order = CONSTANT
      family = MONOMIAL
    []
    # [temperature]
    #   order = CONSTANT
    #   family = MONOMIAL
    # []
  []

  [AuxKernels]
    [enthalpy_water]
      type = PorousFlowPropertyAux
      variable = enthalpy_water
      property = enthalpy
      phase = 0
      execute_on = 'initial timestep_end'
    []
    [enthalpy_gas]
      type = PorousFlowPropertyAux
      variable = enthalpy_gas
      property = enthalpy
      phase = 1
      execute_on = 'initial timestep_end'
    []
    [pressure_water]
      type = PorousFlowPropertyAux
      variable = pressure_water
      property = pressure
      phase = 0
      execute_on = 'initial timestep_end'
    []
    [pressure_gas]
      type = PorousFlowPropertyAux
      variable = pressure_gas
      property = pressure
      phase = 1
      execute_on = 'initial timestep_end'
    []
    [saturation_water]
      type = PorousFlowPropertyAux
      variable = saturation_water
      property = saturation
      phase = 0
      execute_on = 'initial timestep_end'
    []
    [saturation_gas]
      type = PorousFlowPropertyAux
      variable = saturation_gas
      property = saturation
      phase = 1
      execute_on = 'initial timestep_end'
    []
    [density_water]
      type = PorousFlowPropertyAux
      variable = density_water
      property = density
      phase = 0
      execute_on = 'initial timestep_end'
    []
    [density_gas]
      type = PorousFlowPropertyAux
      variable = density_gas
      property = density
      phase = 1
      execute_on = 'initial timestep_end'
    []
    [viscosity_water]
      type = PorousFlowPropertyAux
      variable = viscosity_water
      property = viscosity
      phase = 0
      execute_on = 'initial timestep_end'
    []
    [viscosity_gas]
      type = PorousFlowPropertyAux
      variable = viscosity_gas
      property = viscosity
      phase = 1
      execute_on = 'initial timestep_end'
    []
  []

  [UserObjects]
    [dictator]
      type = PorousFlowDictator
      porous_flow_vars = 'pliquid h'
      number_fluid_phases = 2
      number_fluid_components = 1
    []
    [pc] #from porousflow/test/tests/fluidstate/watervapor.i
      type = PorousFlowCapillaryPressureBC
      pe = 1e5
      lambda = 2
      pc_max = 1e6
    []
    [fs]
      type = PorousFlowWaterVapor
      water_fp = water_tab
      capillary_pressure = pc
    []
  []


  [Variables]
    [pliquid]
      initial_condition = 1e6
    []
    [h]
      initial_condition = 8e4 # if you change it to 8e5 like in water_vapor_tab.i it works fine
    []
  []


  [Kernels]
    [mass]
      type = PorousFlowMassTimeDerivative
      variable = pliquid
    []
    [heat]
      type = PorousFlowEnergyTimeDerivative
      variable = h

    []
  []


  [FluidProperties]
    [water97]
      type = Water97FluidProperties    # IAPWS-IF97
    []
    [water_tab]
      type = TabulatedBicubicFluidProperties
      fp = water97
      fluid_property_file = 'water_extended.csv'
      p_h_variables = true
    []

  []

  [Materials]
      [watervapor]
        type = PorousFlowFluidStateSingleComponent
        porepressure = pliquid
        enthalpy = h
        capillary_pressure = pc
        fluid_state = fs
        temperature_unit = Kelvin
      []

      [porosity_wallrock]
        type = PorousFlowPorosityConst
        porosity = 0.1
      []
      [permeability]
        type = PorousFlowPermeabilityConst
        permeability = '1E-13 0 0   0 1E-13 0   0 0 1E-13'
      []
      [relperm_water] # from watervapor.i
        type = PorousFlowRelativePermeabilityCorey
        n = 2
        phase = 0
      []
      [relperm_gas]  # from watervapor.i
        type = PorousFlowRelativePermeabilityCorey
        n = 3
        phase = 1
      []
      [internal_energy]
        type = PorousFlowMatrixInternalEnergy
        density = 2500 # kg/m^3
        specific_heat_capacity = 1200 # J/kg/K
      []

  []

  [Preconditioning]
    [smp]
      type = SMP
      full = true
      petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
      petsc_options_value = ' lu       mumps'
    []
  []

  [Executioner]
    type = Transient
    solve_type = NEWTON
    end_time = 100
  []

```

GiudGiud commented:
so you forwarded s_from_h_p to Water97, but is it defined there?
Is it defined as s_from_p_h by any chance?

aikubo commented:
> so you forwarded s_from_h_p to Water97, but is it defined there?
It isn't defined in Water97 and I can add it but I find it odd because I'm not sure why this error occurs. 

> Is it defined as s_from_p_h by any chance?
no it is not but it would good to check that

GiudGiud commented:
the backtrace explains the story:
-something in porousflow water calls T_from_p_h
- it s not implemented in TFP, so it goes to the implementation in SinglePhaseFP
- this implementation uses s_from_p_h which is not implemented anywhere

so two solutions:
- implement s_from_p_h (hard imo)
- implement T_from_p_h (easier? you may need to roll out the Newton inversion routine on h_from_p_T)

aikubo committed:
Add s_from_h_p_template functions to
Water97FluidProperties class
add construct flags to _p_h_variables user flag

aikubo commented:
Could you just implement it as basically a copy of `s_from_p_T`? Or is more complexity necessary?

```
template <typename T>
T
Water97FluidProperties::s_from_h_p_template(const T & enthalpy, const T & pressure) const
{
  T temperature = T_from_h_p(enthalpy, pressure);
  return s_from_p_T_template(pressure, temperature);
}

template <typename T>
void
Water97FluidProperties::s_from_h_p_template(
    const T & enthalpy, const T & pressure, T & s, T & ds_dh, T & ds_dp) const
{
  auto functor = [this](const auto & enthalpy, const auto & pressure)
  { return this->s_from_p_T_template(enthalpy, pressure); };

  xyDerivatives(enthalpy, pressure, s, ds_dh, ds_dp, functor);
}

```

And another question, looking closely there's `T_from_p_h` and `T_from_h_p` in Water97FluidProperties (one of which I added)  
And in SinglePhaseFluidProperties there's both. But only `T_from_h_p` in Tabulated Fluid Properties. Aren't these the same? Which should it be?

GiudGiud commented:
but if T_from_h_p is implemented, T_from_p_h is definitely the faster and easier to implement

aikubo commented:
I've added `s_from_h_p` and it works fine but I kept getting issues with 

```
*** ERROR ***
/home/akh/myprojects/moose_projects/dikes/MVE_s_from_h_p_error.i:183.5:
The following error occurred in the FluidProperties 'water_tab' of type TabulatedBicubicFluidProperties.

You must construct pT from vh tables when calling e_from_v_h.
```

When you turn both `construct_pT_from_ve=true` and `construct_pT_from_vh=true` everything works fine. This only occurs in certain regions of the phase diagram for p,h variables in porousflow. I think it might make sense to turn on those flags automatically as part of `p_h_variables` flag in TabulatedFluidProperties. Or if not to note it somewhere.

GiudGiud commented:
> . I think it might make sense to turn on those flags automatically as part of p_h_variables flag in TabulatedFluidProperties. 

When does this flag get set? By the user? If so then sure. 

> Or if not to note it somewhere.

That's fine too.
I'm guessing that you are hitting these 'new' calls because:
- the TFP isnt implementing something, calls Water97, which uses routines that _from_h_p back from the TFP, which are implemented in the TFP but only if the flag is on

aikubo committed:
update to be compatitble with PorousFlowWaterVapor, TabulatedFluidProperties, and Water97FluidProperties with ADReal, FPADReal, and Real types.

aikubo commented:
I was having a hard time tracking down some of the issues I was having and I'll do my best to summarize 

I tried a input file in the Liquid state and it wasn't running at all (MVE below). I added the `s_from_h_p` to Water97FluidProperties and accordingly to TabulatedFluidProperties. 
After that the input file ran but the temperature was wrong. It gave a temperature of 299 K when it should be 377 K. I tracked this to PorousFlowWaterVapor calling `T_from_p_h` but the AD version which wasn't implemented in TabulatedFluidProperties, only the `Real` version. 

I also had to add a `ADReal T_from_p_h` version to WaterFluidProperties because the current version is `FPADReal T_from_p_h_ad`.  

The way I implemented `ADReal T_from_p_h` is imho terrible but I had a hard time compiling between the `FPADReal` and `ADReal` types. It's unclear to me when to use `FPADReal`, it seems to be used mostly for internal FP code. And as far as I can tell the `T_from_p_h` function is only directly called outside the FluidProperties module in `PorousFlowWaterVapor` .  What is the correct usage and correct way to convert between types? 

Finally, I cleaned up the commit history. 

Thanks for your patience bearing with a MOOSE and C++ novice. 

Here's the liquid state MVE which works now:
```
# at 1MPa (1e6 Pa) and 440 kJ/kg (440000) T should be 377 K

[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
  []
[]

[GlobalParams]
  PorousFlowDictator = dictator
[]

[AuxVariables]
  [pressure_gas]
    order = CONSTANT
    family = MONOMIAL
  []
  [pressure_water]
    order = CONSTANT
    family = MONOMIAL
  []
  [enthalpy_gas]
    order = CONSTANT
    family = MONOMIAL
  []
  [enthalpy_water]
    order = CONSTANT
    family = MONOMIAL
  []
  [saturation_gas]
    order = CONSTANT
    family = MONOMIAL
  []
  [saturation_water]
    order = CONSTANT
    family = MONOMIAL
  []
  [density_water]
    order = CONSTANT
    family = MONOMIAL
  []
  [density_gas]
    order = CONSTANT
    family = MONOMIAL
  []
  [viscosity_water]
    order = CONSTANT
    family = MONOMIAL
  []
  [viscosity_gas]
    order = CONSTANT
    family = MONOMIAL
  []
  [temperature]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[AuxKernels]
  [enthalpy_water]
    type = PorousFlowPropertyAux
    variable = enthalpy_water
    property = enthalpy
    phase = 0
    execute_on = 'initial timestep_end'
  []
  [enthalpy_gas]
    type = PorousFlowPropertyAux
    variable = enthalpy_gas
    property = enthalpy
    phase = 1
    execute_on = 'initial timestep_end'
  []
  [pressure_water]
    type = PorousFlowPropertyAux
    variable = pressure_water
    property = pressure
    phase = 0
    execute_on = 'initial timestep_end'
  []
  [pressure_gas]
    type = PorousFlowPropertyAux
    variable = pressure_gas
    property = pressure
    phase = 1
    execute_on = 'initial timestep_end'
  []
  [saturation_water]
    type = PorousFlowPropertyAux
    variable = saturation_water
    property = saturation
    phase = 0
    execute_on = 'initial timestep_end'
  []
  [saturation_gas]
    type = PorousFlowPropertyAux
    variable = saturation_gas
    property = saturation
    phase = 1
    execute_on = 'initial timestep_end'
  []
  [density_water]
    type = PorousFlowPropertyAux
    variable = density_water
    property = density
    phase = 0
    execute_on = 'initial timestep_end'
  []
  [density_gas]
    type = PorousFlowPropertyAux
    variable = density_gas
    property = density
    phase = 1
    execute_on = 'initial timestep_end'
  []
  [viscosity_water]
    type = PorousFlowPropertyAux
    variable = viscosity_water
    property = viscosity
    phase = 0
    execute_on = 'initial timestep_end'
  []
  [viscosity_gas]
    type = PorousFlowPropertyAux
    variable = viscosity_gas
    property = viscosity
    phase = 1
    execute_on = 'initial timestep_end'
  []
  [temperature]
    type = PorousFlowPropertyAux
    variable = temperature
    property = temperature
    execute_on = 'initial timestep_end'
  []
[]

[UserObjects]
  [dictator]
    type = PorousFlowDictator
    porous_flow_vars = 'pliquid h'
    number_fluid_phases = 2
    number_fluid_components = 1
  []
  [pc] #from porousflow/test/tests/fluidstate/watervapor.i
    type = PorousFlowCapillaryPressureBC
    pe = 1e5
    lambda = 2
    pc_max = 1e6
  []
  [fs]
    type = PorousFlowWaterVapor
    water_fp = water_tab
    capillary_pressure = pc
  []
[]

[Variables]
  [pliquid]
    initial_condition = 1e6
  []
  [h]
    initial_condition = 440e3
  []
[]

[Kernels]
  [mass]
    type = PorousFlowMassTimeDerivative
    variable = pliquid
  []
  [heat]
    type = PorousFlowEnergyTimeDerivative
    variable = h
  []
[]

[FluidProperties]
  [water97]
    type = Water97FluidProperties # IAPWS-IF97
  []
  [water_tab]
    type = TabulatedBicubicFluidProperties
    fp = water97
    p_h_variables = true
  []
[]

[Materials]
  [watervapor]
    type = PorousFlowFluidStateSingleComponent
    porepressure = pliquid
    enthalpy = h
    capillary_pressure = pc
    fluid_state = fs
    temperature_unit = Kelvin
  []

  [porosity_wallrock]
    type = PorousFlowPorosityConst
    porosity = 0.1
  []
  [permeability]
    type = PorousFlowPermeabilityConst
    permeability = '1E-13 0 0   0 1E-13 0   0 0 1E-13'
  []
  [relperm_water] # from watervapor.i
    type = PorousFlowRelativePermeabilityCorey
    n = 2
    phase = 0
  []
  [relperm_gas] # from watervapor.i
    type = PorousFlowRelativePermeabilityCorey
    n = 3
    phase = 1
  []
  [internal_energy]
    type = PorousFlowMatrixInternalEnergy
    density = 2500 # kg/m^3
    specific_heat_capacity = 1200 # J/kg/K
  []
[]

[Executioner]
  type = Transient
  solve_type = NEWTON
  end_time = 1
  nl_abs_tol = 1e-12
[]

[Postprocessors]
  [density_water]
    type = ElementAverageValue
    variable = density_water
    execute_on = 'initial timestep_end'
  []
  [density_gas]
    type = ElementAverageValue
    variable = density_gas
    execute_on = 'initial timestep_end'
  []
  [viscosity_water]
    type = ElementAverageValue
    variable = viscosity_water
    execute_on = 'initial timestep_end'
  []
  [viscosity_gas]
    type = ElementAverageValue
    variable = viscosity_gas
    execute_on = 'initial timestep_end'
  []
  [enthalpy_water]
    type = ElementAverageValue
    variable = enthalpy_water
    execute_on = 'initial timestep_end'
  []
  [enthalpy_gas]
    type = ElementAverageValue
    variable = enthalpy_gas
    execute_on = 'initial timestep_end'
  []
  [sg]
    type = ElementAverageValue
    variable = saturation_gas
    execute_on = 'initial timestep_end'
  []
  [sw]
    type = ElementAverageValue
    variable = saturation_water
    execute_on = 'initial timestep_end'
  []
  [pwater]
    type = ElementAverageValue
    variable = pressure_water
    execute_on = 'initial timestep_end'
  []
  [pgas]
    type = ElementAverageValue
    variable = pressure_gas
    execute_on = 'initial timestep_end'
  []
  [temperature]
    type = ElementAverageValue
    variable = temperature
    execute_on = 'initial timestep_end'
  []
  [enthalpy]
    type = ElementAverageValue
    variable = h
    execute_on = 'initial timestep_end'
  []
  [liquid_mass]
    type = PorousFlowFluidMass
    phase = 0
    execute_on = 'initial timestep_end'
  []
  [vapor_mass]
    type = PorousFlowFluidMass
    phase = 1
    execute_on = 'initial timestep_end'
  []
[]

[Outputs]
  file_base = water_vapor_twophase_tab
  csv = true
[]
```

GiudGiud commented:
Sounds good. I ll check this out next week. Then I will be on travel so I may not be able to help this through. If you mind the delay we will have to bring in another reviewer. 

I don't think we need to use the FP version of the ADReal. I get by with using ADReal everywhere

GiudGiud commented:
Looks like this flew under my radar post honeymoon. 

@aikubo do you still need this?

aikubo commented:
Hi @GiudGiud 

Yes I think this would still me useful to me and a few others, perhaps https://github.com/idaholab/moose/discussions/28066 as well.

GiudGiud commented:
https://github.com/idaholab/moose/discussions/28066 was just a convergence issue.

Ok once my own TFP PR is merged (it will conflict) I ll take over this one to get it past the finishing line. Thanks for getting back in touch

GiudGiud committed:
Remove more use of FPADReal
Make sure not to loose derivatives in T_from_p_h ad version for water97

GiudGiud committed:
Add unit tests for non-trivial additions to Water97

GiudGiud committed:
Add unit tests for new routines in TFPs

GiudGiud commented:
ok I rebased and added tests. Machine is borked so I cant run them locally, let's see how civet fares.
@aikubo I did remove the restriction to p_h variables. If you still want that you'll need to make a separate PR

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2377492/) on 422c3f1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27506/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27506/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e0de6598545e9ba8ea902d1ef1ef2560b05decc2`

GiudGiud commented:
@aikubo can I please get some guidance on that test?
The pressures in the CSV file are between 1.9 and 2.3e7 but the IC is at 1e6. 
Should I just raise the pressures? Or do you have an extended tabulation file I can use?

aikubo commented:
Hi @GiudGiud, thanks for working on this. 

Raising the pressures will not show the same behavior since it will change to a different phase region. 

I have been using this script to generate tabulated ranges based on https://github.com/idaholab/moose/discussions/21038#discussioncomment-2779971


```
import CoolProp.CoolProp as cp
import pandas as pd
import numpy as np

# Range of pressure and temperature
minPressure = 1e4
maxPressure = 1.8e9
minTemperature = 275
maxTemperature = 1300

pressure = np.concatenate( (np.linspace(minPressure, 1e5, 5), np.linspace(1.1e5, 1e8, 100), np.linspace(1.1e8, 1e10, 5)))
temperature = np.concatenate( (np.linspace(minTemperature-100, minTemperature, 5), np.arange(minTemperature+5, maxTemperature, 100), np.linspace(maxTemperature+100, maxTemperature+500, 5)))

data = np.zeros((pressure.size * temperature.size, 6))

for i in np.arange(pressure.size):
    for j in np.arange(temperature.size):
        data[i * temperature.size + j, 0] = pressure[i]
        data[i * temperature.size + j, 1] = temperature[j]

        if pressure[i] < minPressure or temperature[j] < minTemperature:
            data[i * temperature.size + j, 0] = pressure[i]
            data[i * temperature.size + j, 1] = temperature[j]
            data[i * temperature.size + j, 2] = cp.PropsSI('D', 'P', minPressure, 'T', minTemperature, 'Water')
            data[i * temperature.size + j, 3] = cp.PropsSI('V', 'P', minPressure, 'T', minTemperature, 'Water')
            data[i * temperature.size + j, 4] = cp.PropsSI('H', 'P', minPressure, 'T', minTemperature, 'Water')
            data[i * temperature.size + j, 5] = cp.PropsSI('U', 'P', minPressure, 'T', minTemperature, 'Water')
            continue
        elif pressure[i] > maxPressure or temperature[j] > maxTemperature:
            data[i * temperature.size + j, 0] = pressure[i]
            data[i * temperature.size + j, 1] = temperature[j]
            data[i * temperature.size + j, 2] = cp.PropsSI('D', 'P', maxPressure, 'T', maxTemperature, 'Water')
            data[i * temperature.size + j, 3] = cp.PropsSI('V', 'P', maxPressure, 'T', maxTemperature, 'Water')
            data[i * temperature.size + j, 4] = cp.PropsSI('H', 'P', maxPressure, 'T', maxTemperature, 'Water')
            data[i * temperature.size + j, 5] = cp.PropsSI('U', 'P', maxPressure, 'T', maxTemperature, 'Water')
            continue
        elif pressure[i] > minPressure and pressure[i] < maxPressure and temperature[j] > minTemperature and temperature[j] < maxTemperature:
            data[i * temperature.size + j, 2] = cp.PropsSI('D', 'P', pressure[i], 'T', temperature[j], 'Water')
            data[i * temperature.size + j, 3] = cp.PropsSI('V', 'P', pressure[i], 'T', temperature[j], 'Water')
            data[i * temperature.size + j, 4] = cp.PropsSI('H', 'P', pressure[i], 'T', temperature[j], 'Water')
            data[i * temperature.size + j, 5] = cp.PropsSI('U', 'P', pressure[i], 'T', temperature[j], 'Water')

# Write out fluid_properties_extended.csv
df = pd.DataFrame(data, columns = ['pressure', 'temperature', 'density', 'viscosity', 'enthalpy', 'internal_energy'])
df.to_csv('water_extended.csv', index=False)
```

I'll push the extended tabulation file.

Commit:
Add extended fluid_properties file for water_vapor_tab test

GiudGiud committed:
Gold porous flow test showing use of tfp for 2 phase vapor

GiudGiud committed:
Remove tests for unimplemented regions
Loosen unit tests that use a reference created for another variable set
Fix a derivative error in waterFP

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2378303/) on 592f8fb wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>e0de65</th>
      <th colspan=4>#27506 592f8f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.93%</td>
      <td>85.03%</td>
      <td>+0.10%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>104783</td>
      <td>104904</td>
      <td>+121</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18593</td>
      <td>18472</td>
      <td>-121</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27506/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27506/coverage/framework)


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>e0de65</th>
      <th colspan=4>#27506 592f8f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.34%</td>
      <td>84.93%</td>
      <td>-0.41%</td>
      <td>56.64%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7610</td>
      <td>7641</td>
      <td>+31</td>
      <td>64</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1307</td>
      <td>1356</td>
      <td>+49</td>
      <td>49</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27506/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27506/coverage/fluid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27506/coverage/xfem)

</details>

## Warnings

- `fluid_properties` new line coverage rate 56.64% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
@joshuahansel for the independent review please

joshuahansel commented on modules/porous_flow/test/tests/fluidstate/tests:
```diff
@@ -181,6 +181,15 @@
+    type = 'CSVDiff'
+    input = 'water_vapor_tab.i'
+    csvdiff = 'water_vapor_twophase_tab.csv'
+    threading = '!pthreads'
```
Why/when do we skip pthreads? I see a lot of "!pthreads" in the test suite but no reason ever listed. We'd like to note reasons as a comment on the line.

GiudGiud replied:
it's tradition for porous_flow module
I ll address that in a follow up

joshuahansel commented on modules/porous_flow/test/tests/fluidstate/fluid_properties_extended.csv:
```diff
@@ -0,0 +1,2311 @@
+10000.0,200.0,999.8920764856763,0.0016821184304554088,7769.115321732408,7759.114242437768
+10000.0,225.0,999.8920764856763,0.0016821184304554088,7769.115321732408,7759.114242437768
+10000.0,250.0,999.8920764856763,0.0016821184304554088,7769.115321732408,7759.114242437768
+10000.0,275.0,0.0,0.0,0.0,0.0
```
Are these 0 lines expected?

GiudGiud replied:
@aikubo 

TFP handles them without complaint :D

joshuahansel commented on modules/porous_flow/test/tests/fluidstate/water_vapor_tab.i:
```diff
@@ -0,0 +1,316 @@
+    variable = pressure_gas
+    execute_on = 'initial timestep_end'
+  []
+  # [temperature]
```
Why is temperature commented out?

GiudGiud replied:
uncommented now

joshuahansel commented on modules/porous_flow/test/tests/fluidstate/water_vapor_tab.i:
```diff
@@ -0,0 +1,316 @@
+  []
+[]
+
+[Executioner]
+  type = Transient
+  solve_type = NEWTON
+  end_time = 1
+  nl_abs_tol = 1e-12
+[]
```
```suggestion
[Problem]
  solve = false
[]

[Executioner]
  type = Steady
[]
```

joshuahansel replied:
And then update your gold file: it should have only one line instead of two now.

GiudGiud replied:
still has two
I think we'd need to pass some execute_on to the csv to prevent any other execution?

joshuahansel replied:
Yeah, I think it'll need `execute_on = INITIAL` in the `Outputs` block.

GiudGiud replied:
ok i ll add it

GiudGiud replied:
done

joshuahansel reviewed (changes requested):
Side note: I never noticed that there was a `vaporPressure` method in `SinglePhaseFluidProperties`. It would have been preferable to use a `TwoPhaseFluidProperties` to meet this need, but maybe this was done because someone didn't have the other (liquid/vapor) properties.

GiudGiud committed:
Address code review:
- clearer message on routines which are forwarded
- simpler test without a solve

Co-authored-by: joshuahansel <joshua.hansel@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2379910/) on 747cad9 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27506/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27506/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e0de6598545e9ba8ea902d1ef1ef2560b05decc2`

GiudGiud commented:
Thank you for the contribution @aikubo !
I m happy TFP can support porous flow needs as well. And this made me understand a lot more the fluid property needs of porous flow. I think we want to support a whole (p, h) variable set for this module, which we are far from

GiudGiud commented on modules/fluid_properties/src/fluidproperties/Water97FluidProperties.C:
```diff
@@ -446,6 +446,38 @@ Water97FluidProperties::s_from_p_T(
+  ADReal T = T_from_p_h_ad(p, h);
+  ADReal entropy = s_from_p_T_template(p, T);
+
+  ds_dh = entropy.derivatives()[1];
+  ds_dp = entropy.derivatives()[0];
```
@aikubo notice I flipped those. if you were getting convergence issues in your model this might have been it
